### DISCUSSION TITLE
Pawn possession events

### DISCUSSION THREAD

**Discussion by riddlemd** [nxrighthere/UnrealCLR#82] [Help] [answered]
I'm not seeing a way for the C# classes to detect events triggered on an Actor and perform logic based on that.

I'd like to be able to be able to assign to an event handler for events like in blueprints; for stuff like Possessed, Unpossessed, etc.

Also, sorta related -- is there a way for a blueprint to call a non static public method on a class that implements it in c#?

So if I have something like:
`var player = new DefaultPlayerCharacter("Player", Blueprint.Load("/Game/Actors/Pawns/Characters/DefaultPlayerCharacterBP"));`

I want to be able to in blueprint graph of DefaultPlayerCharacterBP to be able to send a message directly to DefaultPlayerCharacter, this would be a work around for lack of event support directly in c#.

**Comment by nxrighthere** [comment]
Please, point me to the native equivalent in Unreal Engine API for that.

**Reply by riddlemd** [reply]
From what I can tell the C++ code really sucks at event detection, as event handlers are not really implemented... Might not be doable form a plugin -- which really sucks, because c# has great event support.

I could make a static class that acts as a proxy and routes events through it to the corresponding c# class, but the current managed function calling code needs to support arguments or at minimum the ObjectRefrence class needs to be able to be converted to POCO so I can pass a UObject out of the engine that has the corresponding event data.

**Reply by nxrighthere** [reply]
I can expose a function for checking if a pawn is possessed or not if that would help. So you can invoke it on-demand or implement events logic with checking on a per-frame basis.

**Reply by riddlemd** [reply]
That would help for sure.

How hard would it be to pass custom UObjects out of the engine? Something like ObjectRefrence.ToPoco<T>(). That way I could create a custom UObject with the data I want to pass out of the engine. This would avoid having to deal with the managed function call arguments and all the performance implications of that (mentation in #33)

Could even limit support to serializable types if that would make it easier, as I can reference actor ID's in blueprints and pass those instead of actors themselves.

**Reply by riddlemd** [reply]
To add to this thought, UE4 c++ has json support. You could serialize the UObject and the ObjectReference.ToPoco<T>() could literally just be a json deserialization call on the managed side and the unmanaged side could just expose a string to be deserialized.

**Reply by nxrighthere** [reply]
I'll check how it can be done optimally, thanks for the info.

**Comment by nxrighthere** [comment, marked as answer]
[`Pawn.IsControlled`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/Pawn-IsControlled.md) and [`Pawn.IsPlayerControlled`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/Pawn-IsPlayerControlled.md) exposed in https://github.com/nxrighthere/UnrealCLR/commit/f225d3253f11157f40c47e5f8e69b4710106e334.
